File: backend/app/services/archetype_service.py
from datetime import datetime, timezone
from typing import Optional, Any
from sqlalchemy.orm import Session
from sqlalchemy import or_

from app.db.models import (
    IncidentArchetype,
    InventoryArchetype,
    ArchetypeSource,
)


class ArchetypeService:

    # -------------------------------------------------------------------------
    # Incident Archetype CRUD
    # -------------------------------------------------------------------------

    @staticmethod
    def list_incident_archetypes(
        db: Session,
        source: Optional[str] = None,
        search: Optional[str] = None,
    ) -> list[IncidentArchetype]:
        query = db.query(IncidentArchetype)
        if source:
            query = query.filter(IncidentArchetype.source == source)
        if search:
            query = query.filter(
                or_(
                    IncidentArchetype.name.ilike(f"%{search}%"),
                    IncidentArchetype.description.ilike(f"%{search}%"),
                    IncidentArchetype.id.ilike(f"%{search}%"),
                )
            )
        return query.order_by(IncidentArchetype.name).all()

    @staticmethod
    def get_incident_archetype(
        db: Session, archetype_id: str
    ) -> Optional[IncidentArchetype]:
        return (
            db.query(IncidentArchetype)
            .filter(IncidentArchetype.id == archetype_id)
            .first()
        )

    @staticmethod
    def create_incident_archetype(
        db: Session,
        data: dict[str, Any],
        created_by: Optional[str] = None,
    ) -> IncidentArchetype:
        archetype = IncidentArchetype(
            id=data["id"],
            name=data["name"],
            description=data.get("description"),
            source=data.get("source", ArchetypeSource.USER),
            field_schema=data.get("field_schema", []),
            urgency_rules=data.get("urgency_rules", []),
            implications=data.get("implications", {}),
            default_report_urgency=data.get("default_report_urgency"),
            wikidata_id=data.get("wikidata_id"),
            parent_archetype_id=data.get("parent_archetype_id"),
            created_by=created_by,
        )
        db.add(archetype)
        db.commit()
        db.refresh(archetype)
        return archetype

    @staticmethod
    def update_incident_archetype(
        db: Session,
        archetype_id: str,
        data: dict[str, Any],
    ) -> Optional[IncidentArchetype]:
        archetype = ArchetypeService.get_incident_archetype(db, archetype_id)
        if not archetype:
            return None

        for field, value in data.items():
            if value is not None:
                setattr(archetype, field, value)

        archetype.updated_at = datetime.now(timezone.utc)
        archetype.version += 1

        db.commit()
        db.refresh(archetype)
        return archetype

    @staticmethod
    def delete_incident_archetype(
        db: Session, archetype_id: str
    ) -> bool:
        archetype = ArchetypeService.get_incident_archetype(db, archetype_id)
        if not archetype:
            return False
        db.delete(archetype)
        db.commit()
        return True

    @staticmethod
    def duplicate_incident_archetype(
        db: Session,
        archetype_id: str,
        new_id: str,
        created_by: Optional[str] = None,
    ) -> Optional[IncidentArchetype]:
        original = ArchetypeService.get_incident_archetype(db, archetype_id)
        if not original:
            return None

        data = {
            "id": new_id,
            "name": f"{original.name} (Copy)",
            "description": original.description,
            "source": ArchetypeSource.USER,
            "field_schema": original.field_schema,
            "urgency_rules": original.urgency_rules,
            "implications": original.implications,
            "default_report_urgency": original.default_report_urgency,
            "wikidata_id": original.wikidata_id,
            "parent_archetype_id": archetype_id,
        }
        return ArchetypeService.create_incident_archetype(db, data, created_by)

    # -------------------------------------------------------------------------
    # Inventory Archetype CRUD
    # -------------------------------------------------------------------------

    @staticmethod
    def list_inventory_archetypes(
        db: Session,
        category: Optional[str] = None,
        source: Optional[str] = None,
        search: Optional[str] = None,
    ) -> list[InventoryArchetype]:
        query = db.query(InventoryArchetype)
        if category:
            query = query.filter(InventoryArchetype.category == category)
        if source:
            query = query.filter(InventoryArchetype.source == source)
        if search:
            query = query.filter(
                or_(
                    InventoryArchetype.name.ilike(f"%{search}%"),
                    InventoryArchetype.description.ilike(f"%{search}%"),
                    InventoryArchetype.id.ilike(f"%{search}%"),
                )
            )
        return query.order_by(InventoryArchetype.name).all()

    @staticmethod
    def get_inventory_archetype(
        db: Session, archetype_id: str
    ) -> Optional[InventoryArchetype]:
        return (
            db.query(InventoryArchetype)
            .filter(InventoryArchetype.id == archetype_id)
            .first()
        )

    @staticmethod
    def create_inventory_archetype(
        db: Session,
        data: dict[str, Any],
        created_by: Optional[str] = None,
    ) -> InventoryArchetype:
        archetype = InventoryArchetype(
            id=data["id"],
            name=data["name"],
            description=data.get("description"),
            category=data["category"],
            source=data.get("source", ArchetypeSource.USER),
            field_schema=data.get("field_schema", []),
            urgency_rules=data.get("urgency_rules", []),
            resolves_needs=data.get("resolves_needs", []),
            target_demographics=data.get("target_demographics", []),
            physical_properties=data.get("physical_properties"),
            quantity_unit=data["quantity_unit"],
            food_properties=data.get("food_properties"),
            medical_properties=data.get("medical_properties"),
            brand=data.get("brand"),
            learning=data.get("learning"),
            wikidata_id=data.get("wikidata_id"),
            parent_archetype_id=data.get("parent_archetype_id"),
            created_by=created_by,
        )
        db.add(archetype)
        db.commit()
        db.refresh(archetype)
        return archetype

    @staticmethod
    def update_inventory_archetype(
        db: Session,
        archetype_id: str,
        data: dict[str, Any],
    ) -> Optional[InventoryArchetype]:
        archetype = ArchetypeService.get_inventory_archetype(db, archetype_id)
        if not archetype:
            return None

        for field, value in data.items():
            if value is not None:
                setattr(archetype, field, value)

        archetype.updated_at = datetime.now(timezone.utc)
        archetype.version += 1

        db.commit()
        db.refresh(archetype)
        return archetype

    @staticmethod
    def delete_inventory_archetype(
        db: Session, archetype_id: str
    ) -> bool:
        archetype = ArchetypeService.get_inventory_archetype(db, archetype_id)
        if not archetype:
            return False
        db.delete(archetype)
        db.commit()
        return True

    @staticmethod
    def duplicate_inventory_archetype(
        db: Session,
        archetype_id: str,
        new_id: str,
        created_by: Optional[str] = None,
    ) -> Optional[InventoryArchetype]:
        original = ArchetypeService.get_inventory_archetype(db, archetype_id)
        if not original:
            return None

        data = {
            "id": new_id,
            "name": f"{original.name} (Copy)",
            "description": original.description,
            "category": original.category,
            "source": ArchetypeSource.USER,
            "field_schema": original.field_schema,
            "urgency_rules": original.urgency_rules,
            "resolves_needs": original.resolves_needs,
            "target_demographics": original.target_demographics,
            "physical_properties": original.physical_properties,
            "quantity_unit": original.quantity_unit,
            "food_properties": original.food_properties,
            "medical_properties": original.medical_properties,
            "brand": original.brand,
            "learning": original.learning,
            "wikidata_id": original.wikidata_id,
            "parent_archetype_id": archetype_id,
        }
        return ArchetypeService.create_inventory_archetype(db, data, created_by)

    # -------------------------------------------------------------------------
    # Urgency Calculation
    # -------------------------------------------------------------------------

    @staticmethod
    def calculate_urgency(
        urgency_rules: list[dict[str, Any]],
        values: dict[str, Any],
        default_urgency: Optional[str] = None,
    ) -> dict[str, str]:
        for rule in urgency_rules:
            field = rule.get("field")
            operator = rule.get("operator")
            rule_value = rule.get("value")
            set_urgency = rule.get("setUrgency")
            message = rule.get("message", "")

            if field not in values:
                continue

            actual_value = values[field]

            if ArchetypeService._evaluate_operator(actual_value, operator, rule_value):
                return {
                    "urgency": set_urgency,
                    "reason": message,
                }

        return {
            "urgency": default_urgency or "medium",
            "reason": "Default urgency",
        }

    @staticmethod
    def _evaluate_operator(
        actual: Any, operator: str, expected: Any
    ) -> bool:
        try:
            if operator == "<":
                return actual < expected
            elif operator == ">":
                return actual > expected
            elif operator == "=":
                return actual == expected
            elif operator == "<=":
                return actual <= expected
            elif operator == ">=":
                return actual >= expected
        except TypeError:
            return False
        return False

    # -------------------------------------------------------------------------
    # Override / Parent Detection
    # -------------------------------------------------------------------------

    @staticmethod
    def is_override(archetype: IncidentArchetype | InventoryArchetype) -> bool:
        return archetype.parent_archetype_id is not None

    @staticmethod
    def get_base_incident_archetype(
        db: Session, archetype: IncidentArchetype
    ) -> IncidentArchetype:
        while archetype.parent_archetype_id:
            parent = (
                db.query(IncidentArchetype)
                .filter(IncidentArchetype.id == archetype.parent_archetype_id)
                .first()
            )
            if not parent:
                break
            archetype = parent
        return archetype

    @staticmethod
    def get_base_inventory_archetype(
        db: Session, archetype: InventoryArchetype
    ) -> InventoryArchetype:
        while archetype.parent_archetype_id:
            parent = (
                db.query(InventoryArchetype)
                .filter(InventoryArchetype.id == archetype.parent_archetype_id)
                .first()
            )
            if not parent:
                break
            archetype = parent
        return archetype

    @staticmethod
    def get_overrides_incident(
        db: Session, archetype_id: str
    ) -> list[IncidentArchetype]:
        archetype = ArchetypeService.get_incident_archetype(db, archetype_id)
        if not archetype:
            return []
        return archetype.children or []

    @staticmethod
    def get_overrides_inventory(
        db: Session, archetype_id: str
    ) -> list[InventoryArchetype]:
        archetype = ArchetypeService.get_inventory_archetype(db, archetype_id)
        if not archetype:
            return []
        return archetype.children or []

    # -------------------------------------------------------------------------
    # First-to-Enter Hook (placeholder for Wikidata integration)
    # -------------------------------------------------------------------------

    @staticmethod
    def find_archetype_by_name_incident(
        db: Session, name: str
    ) -> Optional[IncidentArchetype]:
        return (
            db.query(IncidentArchetype)
            .filter(IncidentArchetype.name.ilike(name))
            .first()
        )

    @staticmethod
    def find_archetype_by_name_inventory(
        db: Session, name: str
    ) -> Optional[InventoryArchetype]:
        return (
            db.query(InventoryArchetype)
            .filter(InventoryArchetype.name.ilike(name))
            .first()
        )

File: frontend/src/features/archetypes/services/archetypeService.ts
import * as archetypesApi from "../../../shared/api/archetypes";
import type {
  IncidentArchetype,
  InventoryArchetype,
  ArchetypeListEntry,
  UrgencyLevel,
} from "../types/archetypes.types";
import type {
  ApiIncidentArchetypeResponse,
  ApiInventoryArchetypeResponse,
  ApiArchetypeListEntry,
  ApiIncidentArchetypeCreate,
  ApiInventoryArchetypeCreate,
  ApiIncidentArchetypeUpdate,
  ApiInventoryArchetypeUpdate,
} from "../../../shared/api/types";

function mapFieldSchema(api: ApiIncidentArchetypeResponse["field_schema"]): IncidentArchetype["fieldSchema"] {
  return api.map((f) => ({
    field: f.field,
    label: f.label,
    type: f.type,
    required: f.required,
    options: f.options,
    defaultValue: f.defaultValue,
  }));
}

function mapUrgencyRules(api: ApiIncidentArchetypeResponse["urgency_rules"]): IncidentArchetype["urgencyRules"] {
  return api.map((r) => ({
    field: r.field,
    operator: r.operator,
    value: r.value,
    setUrgency: r.setUrgency,
    message: r.message,
  }));
}

function mapListEntry(api: ApiArchetypeListEntry): ArchetypeListEntry {
  return {
    type: api.type,
    id: api.id,
    name: api.name,
    category: api.category,
    source: api.source,
    version: api.version,
    parentArchetypeId: api.parent_archetype_id,
  };
}

function mapIncidentArchetype(api: ApiIncidentArchetypeResponse): IncidentArchetype {
  return {
    id: api.id,
    name: api.name,
    description: api.description,
    category: "incident",
    source: api.source,
    fieldSchema: mapFieldSchema(api.field_schema),
    urgencyRules: mapUrgencyRules(api.urgency_rules),
    implications: api.implications,
    defaultReportUrgency: api.default_report_urgency as UrgencyLevel | undefined,
    wikidataId: api.wikidata_id,
    parentArchetypeId: api.parent_archetype_id,
    createdBy: api.created_by,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
    version: api.version,
  };
}

function mapInventoryArchetype(api: ApiInventoryArchetypeResponse): InventoryArchetype {
  return {
    id: api.id,
    name: api.name,
    description: api.description,
    category: api.category,
    source: api.source,
    fieldSchema: mapFieldSchema(api.field_schema),
    urgencyRules: mapUrgencyRules(api.urgency_rules),
    resolvesNeeds: api.resolves_needs,
    targetDemographics: api.target_demographics,
    physicalProperties: api.physical_properties,
    quantityUnit: api.quantity_unit,
    foodProperties: api.food_properties,
    medicalProperties: api.medical_properties,
    brand: api.brand,
    learning: api.learning,
    wikidataId: api.wikidata_id,
    parentArchetypeId: api.parent_archetype_id,
    createdBy: api.created_by,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
    version: api.version,
  };
}

export async function fetchArchetypeList(params?: {
  category?: string;
  source?: string;
  search?: string;
}): Promise<ArchetypeListEntry[]> {
  const entries = await archetypesApi.listArchetypes(params as Parameters<typeof archetypesApi.listArchetypes>[0]);
  return entries.map(mapListEntry);
}

export async function fetchIncidentArchetype(id: string): Promise<IncidentArchetype> {
  const api = await archetypesApi.getIncidentArchetype(id);
  return mapIncidentArchetype(api);
}

export async function fetchInventoryArchetype(id: string): Promise<InventoryArchetype> {
  const api = await archetypesApi.getInventoryArchetype(id);
  return mapInventoryArchetype(api);
}

export async function createIncidentArchetype(
  data: Omit<IncidentArchetype, "source" | "createdBy" | "createdAt" | "updatedAt" | "version" | "parentArchetypeId" | "wikidataId">
): Promise<IncidentArchetype> {
  const apiData: ApiIncidentArchetypeCreate = {
    id: data.id,
    name: data.name,
    description: data.description,
    field_schema: data.fieldSchema.map((f) => ({
      field: f.field,
      label: f.label,
      type: f.type,
      required: f.required,
      options: f.options,
      defaultValue: f.defaultValue,
    })),
    urgency_rules: data.urgencyRules.map((r) => ({
      field: r.field,
      operator: r.operator,
      value: r.value,
      setUrgency: r.setUrgency,
      message: r.message,
    })),
    implications: data.implications,
    default_report_urgency: data.defaultReportUrgency,
  };
  const api = await archetypesApi.createIncidentArchetype(apiData);
  return mapIncidentArchetype(api);
}

export async function createInventoryArchetype(
  data: Omit<InventoryArchetype, "source" | "createdBy" | "createdAt" | "updatedAt" | "version" | "parentArchetypeId" | "wikidataId">
): Promise<InventoryArchetype> {
  const apiData: ApiInventoryArchetypeCreate = {
    id: data.id,
    name: data.name,
    description: data.description,
    category: data.category,
    field_schema: data.fieldSchema.map((f) => ({
      field: f.field,
      label: f.label,
      type: f.type,
      required: f.required,
      options: f.options,
      defaultValue: f.defaultValue,
    })),
    urgency_rules: data.urgencyRules.map((r) => ({
      field: r.field,
      operator: r.operator,
      value: r.value,
      setUrgency: r.setUrgency,
      message: r.message,
    })),
    resolves_needs: data.resolvesNeeds,
    target_demographics: data.targetDemographics,
    physical_properties: data.physicalProperties,
    quantity_unit: data.quantityUnit,
    food_properties: data.foodProperties,
    medical_properties: data.medicalProperties,
    brand: data.brand,
    learning: data.learning,
  };
  const api = await archetypesApi.createInventoryArchetype(apiData);
  return mapInventoryArchetype(api);
}

export async function updateIncidentArchetype(
  id: string,
  data: Partial<IncidentArchetype>
): Promise<IncidentArchetype> {
  const apiData: ApiIncidentArchetypeUpdate = {};
  if (data.name !== undefined) apiData.name = data.name;
  if (data.description !== undefined) apiData.description = data.description;
  if (data.fieldSchema) apiData.field_schema = data.fieldSchema.map((f) => ({
    field: f.field,
    label: f.label,
    type: f.type,
    required: f.required,
    options: f.options,
    defaultValue: f.defaultValue,
  }));
  if (data.urgencyRules) apiData.urgency_rules = data.urgencyRules.map((r) => ({
    field: r.field,
    operator: r.operator,
    value: r.value,
    setUrgency: r.setUrgency,
    message: r.message,
  }));
  if (data.implications) apiData.implications = data.implications;
  if (data.defaultReportUrgency) apiData.default_report_urgency = data.defaultReportUrgency;

  const api = await archetypesApi.updateIncidentArchetype(id, apiData);
  return mapIncidentArchetype(api);
}

export async function updateInventoryArchetype(
  id: string,
  data: Partial<InventoryArchetype>
): Promise<InventoryArchetype> {
  const apiData: ApiInventoryArchetypeUpdate = {};
  if (data.name !== undefined) apiData.name = data.name;
  if (data.description !== undefined) apiData.description = data.description;
  if (data.fieldSchema) apiData.field_schema = data.fieldSchema.map((f) => ({
    field: f.field,
    label: f.label,
    type: f.type,
    required: f.required,
    options: f.options,
    defaultValue: f.defaultValue,
  }));
  if (data.urgencyRules) apiData.urgency_rules = data.urgencyRules.map((r) => ({
    field: r.field,
    operator: r.operator,
    value: r.value,
    setUrgency: r.setUrgency,
    message: r.message,
  }));
  if (data.resolvesNeeds) apiData.resolves_needs = data.resolvesNeeds;
  if (data.targetDemographics) apiData.target_demographics = data.targetDemographics;
  if (data.physicalProperties) apiData.physical_properties = data.physicalProperties;
  if (data.quantityUnit) apiData.quantity_unit = data.quantityUnit;
  if (data.foodProperties) apiData.food_properties = data.foodProperties;
  if (data.medicalProperties) apiData.medical_properties = data.medicalProperties;
  if (data.brand) apiData.brand = data.brand;
  if (data.learning) apiData.learning = data.learning;

  const api = await archetypesApi.updateInventoryArchetype(id, apiData);
  return mapInventoryArchetype(api);
}

export async function deleteIncidentArchetype(id: string): Promise<void> {
  await archetypesApi.deleteIncidentArchetype(id);
}

export async function deleteInventoryArchetype(id: string): Promise<void> {
  await archetypesApi.deleteInventoryArchetype(id);
}

export async function duplicateIncidentArchetype(
  id: string,
  newId: string
): Promise<IncidentArchetype> {
  const api = await archetypesApi.duplicateIncidentArchetype(id, newId);
  return mapIncidentArchetype(api);
}

export async function duplicateInventoryArchetype(
  id: string,
  newId: string
): Promise<InventoryArchetype> {
  const api = await archetypesApi.duplicateInventoryArchetype(id, newId);
  return mapInventoryArchetype(api);
}

export async function calculateIncidentUrgency(
  id: string,
  values: Record<string, unknown>
): Promise<{ urgency: string; reason: string }> {
  return archetypesApi.calculateIncidentUrgency(id, values);
}

export async function calculateInventoryArchetypeUrgency(
  id: string,
  values: Record<string, unknown>
): Promise<{ urgency: string; reason: string }> {
  return archetypesApi.calculateInventoryArchetypeUrgency(id, values);
}
